src/models/room.ts

```diff
@@ -1712,45 +1663,44 @@ export class Room extends TypedEventEmitter<EmittedEvents, RoomEventHandlerMap>
             }
         }
 
-        const thread = new Thread(rootEvent, {
+        const thread = new Thread(threadId, rootEvent, {
             initialEvents: events,
             room: this,
             client: this.client,
         });
+
         // If we managed to create a thread and figure out its `id` then we can use it
-        if (thread.id) {
-            this.threads.set(thread.id, thread);
-            this.reEmitter.reEmit(thread, [
-                ThreadEvent.Update,
-                ThreadEvent.NewReply,
-                RoomEvent.Timeline,
-                RoomEvent.TimelineReset,
-            ]);
-
-            if (!this.lastThread || this.lastThread.rootEvent?.localTimestamp < rootEvent?.localTimestamp) {
-                this.lastThread = thread;
-            }
+        this.threads.set(thread.id, thread);
+        this.reEmitter.reEmit(thread, [
+            ThreadEvent.Update,
+            ThreadEvent.NewReply,
+            RoomEvent.Timeline,
+            RoomEvent.TimelineReset,
+        ]);
 
-            this.emit(ThreadEvent.New, thread, toStartOfTimeline);
-
-            if (this.threadsReady) {
-                this.threadsTimelineSets.forEach(timelineSet => {
-                    if (thread.rootEvent) {
-                        if (Thread.hasServerSideSupport) {
-                            timelineSet.addLiveEvent(thread.rootEvent);
-                        } else {
-                            timelineSet.addEventToTimeline(
-                                thread.rootEvent,
-                                timelineSet.getLiveTimeline(),
-                                toStartOfTimeline,
-                            );
-                        }
-                    }
-                });
-            }
+        if (!this.lastThread || this.lastThread.rootEvent?.localTimestamp < rootEvent?.localTimestamp) {
+            this.lastThread = thread;
+        }
+
+        this.emit(ThreadEvent.New, thread, toStartOfTimeline);
 
-            return thread;
+        if (this.threadsReady) {
+            this.threadsTimelineSets.forEach(timelineSet => {
+                if (thread.rootEvent) {
+                    if (Thread.hasServerSideSupport) {
+                        timelineSet.addLiveEvent(thread.rootEvent);
+                    } else {
+                        timelineSet.addEventToTimeline(
+                            thread.rootEvent,
+                            timelineSet.getLiveTimeline(),
+                            toStartOfTimeline,
+                        );
+                    }
+                }
+            });
         }
```

Comment:
Race conditions of events coming in as the thread is being created async during a subsequent /sync response, now threads are always created and stored synchronously (so findEventById continue to function) then more data is fetched into them as soon as possible

Why?

Reply:
> There is no context around why/what besides the code itself.

Its in the backref'd PR

https://github.com/matrix-org/matrix-react-sdk/pull/8448
https://github.com/vector-im/element-web/issues/21627

> I was curious because I saw as a "What's new?" item when updating and checking whether this fixes this sort of problem, https://github.com/matrix-org/element-web-rageshakes/issues/12681

Unlikely, that looks more like https://github.com/vector-im/element-web/issues/21613#issuecomment-1117103189

Reply:
> Its in the backref'd PR
> 
> [matrix-org/matrix-react-sdk#8448](https://github.com/matrix-org/matrix-react-sdk/pull/8448)
> [vector-im/element-web#21627](https://github.com/vector-im/element-web/issues/21627)

ohh, that's confusing just following the same name PR breadcrumbs in the activity log. I added those links to the PR description.

Thanks for the context 🙂

Reply:
Normally we only link from the downstream project upstream, given that js-sdk *shouldn't* be tailoring to a specific consumer

Reply:
That seems a bit backwards.

The usage of `matrix-js-sdk` is where new bugs are found and new usages are desired. Giving examples and explaining that these race conditions can cause reactions not to show on threaded messages because of xxx, is pretty relevant to the changes here.

I think even more context should be explained here besides linking to those issues btw. I don't know what the code is actually doing here but it could explain something like, "Instead of adding events directly to the timeline, we emit `ThreadEvent.Update` which clients can react to and refresh the timeline. This fixes the order so that whatever is not missing in a race when this happens."